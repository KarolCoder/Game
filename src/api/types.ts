export type SwapiUrlType = 'people' | 'starships';

export type PeopleResurcesUri =
  | 'height'
  | 'mass'
  | 'films'
  | 'species'
  | 'starships'
  | 'vehicles';

export type StarshipsResourcesUri =
  | 'cost_in_credits'
  | 'length'
  | 'crew'
  | 'passengers'
  | 'max_atmosphering_speed'
  | 'hyperdrive_rating'
  | 'MGLT'
  | 'cargo_capacity'
  | 'consumables'
  | 'films'
  | 'pilots';

export type Resource<T> = {uriParam: T; message: string};
