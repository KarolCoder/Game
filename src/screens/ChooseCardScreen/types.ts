export type GameResult = 'Tie' | 'Winner' | 'Loser' | 'unknown' | undefined;
