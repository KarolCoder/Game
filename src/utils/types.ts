import {
  PeopleResurcesUri,
  Resource,
  ResultValue,
  StarshipsResourcesUri,
} from '@/api/types';

export type GameCategory = 'Starships' | 'People';
export type GameResource = Resource<StarshipsResourcesUri | PeopleResurcesUri>;
export type GameCard = Resource<StarshipsResourcesUri | PeopleResurcesUri>;
export type WinCounter = {you?: number; opponent?: number};
export type GameResult = 'Tie' | 'Winner' | 'Loser' | 'unknown' | undefined;
export type ResourcesFromData = {
  firstResource?: ResultValue;
  secondResource?: ResultValue;
};

export type GameValues = {
  category?: GameCategory;
  resource?: GameResource;
  selectedCard?: number;
  winCounter?: WinCounter;
  result?: GameResult;
  resourcesFromData?: ResourcesFromData;
};
