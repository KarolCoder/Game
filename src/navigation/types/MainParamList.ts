import {PeopleResurcesUri, Resource, StarshipsResourcesUri} from '@/api/types';

type ChooseCardProps = {
  activeCategory: 'Starships' | 'People';
  activeResource: Resource<StarshipsResourcesUri | PeopleResurcesUri>;
};

export type MainParamList = {
  Welcome: undefined;
  Game: undefined;
  ChooseCard: ChooseCardProps;
};
