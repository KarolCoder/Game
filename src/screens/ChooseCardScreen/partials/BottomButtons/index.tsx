import {Spacer} from '@/components';
import React from 'react';
import {Row} from '../../styles';
import {BottomButton} from './styles';

interface BottomButtonsProps {
  isDisabled?: boolean;
  onLeftPress: () => void;
  onRightPress: () => void;
}

export const BottomButtons: React.FC<BottomButtonsProps> = ({
  isDisabled,
  onLeftPress,
  onRightPress,
}) => {
  return (
    <Row>
      <BottomButton
        mode="contained"
        disabled={isDisabled}
        onPress={onLeftPress}>
        Reset cards
      </BottomButton>
      <Spacer width={24} />
      <BottomButton
        mode="contained"
        disabled={isDisabled}
        onPress={onRightPress}>
        Go to start
      </BottomButton>
    </Row>
  );
};
