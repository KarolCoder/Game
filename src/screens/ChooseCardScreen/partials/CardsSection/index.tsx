import {Spacer} from '@/components';
import React from 'react';
import {Title, useTheme} from 'react-native-paper';
import {Row} from '../../styles';
import {Card, SelectTitle} from './styles';

interface CardsSectionProps {
  setSelectedCard: React.Dispatch<React.SetStateAction<number | undefined>>;
  refreshData: () => void;
  getNextPage: () => Promise<void>;
  firstValueDisplayed?: string;
  secondValueDisplayed?: string;
  resourcesFromData?: {
    firstResource: string | string[] | undefined;
    secondResource: string | string[] | undefined;
  };
  nextPage?: string;
  status?: string;
}

export const CardsSection: React.FC<CardsSectionProps> = ({
  resourcesFromData,
  setSelectedCard,
  refreshData,
  getNextPage,
  firstValueDisplayed,
  secondValueDisplayed,
  nextPage,
  status,
}) => {
  const {colors} = useTheme();
  return (
    <>
      <SelectTitle textColor={colors.primary}>Choose card</SelectTitle>
      <Row>
        <Card
          disabled={
            (!!resourcesFromData?.firstResource &&
              !!resourcesFromData?.secondResource) ||
            status === 'fetching'
          }
          onPress={() => {
            setSelectedCard(0);
            if (!nextPage) {
              refreshData();
            } else {
              getNextPage();
            }
          }}>
          <Title>{firstValueDisplayed ? firstValueDisplayed : '?'}</Title>
        </Card>
        <Spacer width={24} />
        <Card
          disabled={
            (!!resourcesFromData?.firstResource &&
              !!resourcesFromData?.secondResource) ||
            status === 'fetching'
          }
          onPress={() => {
            setSelectedCard(1);
            if (!nextPage) {
              refreshData();
            } else {
              getNextPage();
            }
          }}>
          <Title>{secondValueDisplayed ? secondValueDisplayed : '?'}</Title>
        </Card>
      </Row>
    </>
  );
};
