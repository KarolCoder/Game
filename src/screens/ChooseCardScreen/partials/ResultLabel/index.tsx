import {BigText} from '../../styles';
import React from 'react';
import {useTheme} from 'react-native-paper';
import {GameResult} from '@/utils/types';

interface ResultLabelProps {
  result: GameResult;
}

export const ResultLabel: React.FC<ResultLabelProps> = ({result}) => {
  const {colors} = useTheme();

  return result === 'Winner' ? (
    <BigText textColor={colors.primary}>Congratualtions, you won!!!</BigText>
  ) : result === 'Loser' ? (
    <BigText textColor={colors.primary}>You lose, loser :D</BigText>
  ) : result === 'Tie' ? (
    <BigText textColor={colors.primary}>Tie, try again</BigText>
  ) : !result ? (
    <BigText textColor={colors.primary}>
      Couldn't resolve winner please go to start
    </BigText>
  ) : null;
};
