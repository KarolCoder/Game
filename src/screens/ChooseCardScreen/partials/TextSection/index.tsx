import React from 'react';
import {useTheme} from 'react-native-paper';
import {BigText, Row, SmallText} from '../../styles';
import {CategoryBox} from './styles';

interface TextSectionProps {
  leftTopText?: string;
  rightTopText?: string;
  bottomLeftText?: string | number;
  bottomRightText?: string | number;
}

export const TextSection: React.FC<TextSectionProps> = ({
  leftTopText = '',
  rightTopText = '',
  bottomLeftText = '',
  bottomRightText = '',
}) => {
  const {colors} = useTheme();
  return (
    <Row>
      <CategoryBox>
        <SmallText textColor={colors.primary}>{leftTopText}</SmallText>
        <BigText textColor={colors.primary}>{bottomLeftText}</BigText>
      </CategoryBox>
      <CategoryBox>
        <SmallText textColor={colors.primary}>{rightTopText}</SmallText>
        <BigText textColor={colors.primary}>{bottomRightText}</BigText>
      </CategoryBox>
    </Row>
  );
};
