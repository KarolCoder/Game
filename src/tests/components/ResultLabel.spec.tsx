import {ResultLabel} from '@/screens/ChooseCardScreen/partials/ResultLabel';
import {GameResult} from '@/utils/types';
import {render} from '@testing-library/react-native';
import React from 'react';

import renderer from 'react-test-renderer';

it('result label renders correctly', async () => {
  const resultLabelData: Array<{result: GameResult; text: string}> = [
    {result: 'Winner', text: 'Congratualtions, you won!!!'},
    {result: 'Loser', text: 'You lose, loser :D'},
    {result: 'Tie', text: 'Tie, try again'},
    {result: undefined, text: "Couldn't resolve winner please go to start"},
  ];
  resultLabelData.forEach(({result, text}) => {
    const component = <ResultLabel result={result} />;
    const resultRenderer = renderer.create(component).toJSON();
    const {getByText} = render(component);
    expect(resultRenderer).toMatchSnapshot();
    getByText(text);
  });
});
