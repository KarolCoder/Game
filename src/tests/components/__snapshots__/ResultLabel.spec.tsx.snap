// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`result label renders correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "textAlign": "left",
      },
      Array [
        Object {
          "textAlign": "left",
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "System",
          "fontWeight": "500",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "fontSize": 20,
            "letterSpacing": 0.15,
            "lineHeight": 30,
            "marginVertical": 2,
          },
          Array [
            Object {
              "color": "#6200ee",
            },
          ],
        ],
      ],
    ]
  }
  textColor="#6200ee"
>
  Congratualtions, you won!!!
</Text>
`;

exports[`result label renders correctly 2`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "textAlign": "left",
      },
      Array [
        Object {
          "textAlign": "left",
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "System",
          "fontWeight": "500",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "fontSize": 20,
            "letterSpacing": 0.15,
            "lineHeight": 30,
            "marginVertical": 2,
          },
          Array [
            Object {
              "color": "#6200ee",
            },
          ],
        ],
      ],
    ]
  }
  textColor="#6200ee"
>
  You lose, loser :D
</Text>
`;

exports[`result label renders correctly 3`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "textAlign": "left",
      },
      Array [
        Object {
          "textAlign": "left",
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "System",
          "fontWeight": "500",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "fontSize": 20,
            "letterSpacing": 0.15,
            "lineHeight": 30,
            "marginVertical": 2,
          },
          Array [
            Object {
              "color": "#6200ee",
            },
          ],
        ],
      ],
    ]
  }
  textColor="#6200ee"
>
  Tie, try again
</Text>
`;

exports[`result label renders correctly 4`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      Object {
        "textAlign": "left",
      },
      Array [
        Object {
          "textAlign": "left",
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "System",
          "fontWeight": "500",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "fontSize": 20,
            "letterSpacing": 0.15,
            "lineHeight": 30,
            "marginVertical": 2,
          },
          Array [
            Object {
              "color": "#6200ee",
            },
          ],
        ],
      ],
    ]
  }
  textColor="#6200ee"
>
  Couldn't resolve winner please go to start
</Text>
`;
